module/python/bench_ebpf_space.py

```diff
@@ -0,0 +1,144 @@
+parser.add_argument('-d', action='store', dest='data_size', help="Data Size", default=8, type=int)
+parser.add_argument('-z', action='store_true', dest='zero', help="Data Size", default=False)
+
+def write(data, fd=3):
+    try:
+        os.write(fd, bytes(data, 'ascii'))
+    except OSError:
+        return -1
+
+args = parser.parse_args()
+pid = os.getpid()
+
+txtZero = "false"
+
+if args.zero:
+    txtZero = "true"
+
+txt = f"""
+#include <linux/sched.h>
+
+struct data_ts {{
+    u64 ts;
+}};
+
+struct ShiftXor {{
+    u64 w;
+    u64 x;
+    u64 y;
+    u64 z;
+}};
+
+static u64 simplerand(struct ShiftXor* rand) {{
+    u64 t = rand->x;
+    t ^= t << 11;
+    t ^= t >> 8;
+    rand->x = rand->y;
+    rand->y = rand->z;
+    rand->z = rand->w;
+    rand->w ^= rand->w >> 19;
+    rand->w ^= t;
+    return rand->w;
+}}
+
+struct sizer {{
+    u32 size[{args.data_size}/4];
+}};
+
+BPF_PERF_OUTPUT(events);
+BPF_PERF_OUTPUT(useless);
+
+BPF_ARRAY(array, struct sizer, {args.size});
+
+BPF_ARRAY(on, struct data_ts, 1);
+
+int probe(struct pt_regs *ctx) {{
+    u32 zero_key = 0;
+    struct data_ts* counter = on.lookup(&zero_key);
+    if(counter == NULL) return 0;
+    u64 val = __sync_lock_test_and_set(&(counter->ts), 1);
+    if( val != 0 ) return 0;
+
+    struct data_ts data;
+    struct ShiftXor rand = {{1, 4, 7, 13}};
+    struct sizer whee;
+    __u32 returner = 0;
+
+    struct ShiftXor rand1 = {{1, 4, 7, 13}};
+
+    u64 start = bpf_ktime_get_ns();
+    for(u64 i = 0; i < {args.number}; i++)
+    {{
+        u32 key = simplerand(&rand1) % {args.size};
+        struct sizer* ptr = &whee;
+        if(!{txtZero}) {{
+            ptr = array.lookup(&key);
+            if(!ptr) return -1;
+        }} else {{
+            ptr->size[0] = key;
+        }}
+        returner ^= (*ptr).size[0];
+    }}
+    u64 stop = bpf_ktime_get_ns();
+    data.ts = stop - start;
+
+    events.perf_submit(ctx, &data, sizeof(struct data_ts));
+    useless.perf_submit(ctx, &returner, sizeof(returner));
+    return 0;
+}}
+"""
+
+
+def simplerand(rand: list[int]) -> (int, list[int]):
+    w = rand[0]
+    x = rand[1]
+    y = rand[2]
+    z = rand[3]
+    t = x
+    t ^= t << 11
+    t ^= t >> 8
+    x = y
+    y = z
+    z = w
+    w ^= w >> 19
+    w ^= t
+    return (w, [w, x, y, z])
+
+
+bpf_ctx = BPF(text=txt)
+on = bpf_ctx["on"]
+on[c_int(0)]=c_int(1)
+
+
+array = bpf_ctx["array"]
+
+rand = [1, 4, 7, 13]
+
+for i in range(0, args.size):
+    val, rand = simplerand(rand)
+    array[c_int(i)] = c_int(val)
+
+bpf_ctx.attach_kprobe(event='do_nanosleep', fn_name="probe")
+
+ready = True
+
+def print_event(cpu, data, size):
+    output = bpf_ctx["events"].event(data)
+    write("get_iterations %ld\tmap_size %ld\tvalue_size %d\tTime(ns) %ld\n" % (args.number, args.size, args.data_size, output.ts))
+    return -1
+
+bpf_ctx["events"].open_perf_buffer(print_event)
+
+while ready:
+    on[c_int(0)]=c_int(0)
+    bpf_ctx.perf_buffer_poll()
```

Comment:
Hangs here unless a time.sleep(0.001) call is added before; unclear what else is expected to trigger `do_nanosleep`.

Reply:
`do_nanosleep` is trigger by lots of applications, but there is no harm in specifying one ourselves.